File: go/RightSideview/RightSideview.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(logN)
func rightSideView(root *TreeNode) []int {
	queue := make([][]*TreeNode, 0)
	queue = append(queue, []*TreeNode{root})
	visible := make([]int, 0)

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		next := make([]*TreeNode, 0)

		for len(curr) > 0 {
			node := curr[0]
			curr = curr[1:]

			if node.Left != nil {
				next = append(next, node.Left)
			}

			if node.Right != nil {
				next = append(next, node.Right)
			}
		}

		if len(next) > 0 {
			queue = append(queue, next)
			visible = append(visible, next[len(next)-1].Val)
		}

	}
	return visible
}

File: python/rightSideView.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


# Time -> 0(N) Space -> 0(logN)
def rightSideView(root: TreeNode | None) -> list[int]:
    visible = []
    queue = []
    queue.append([root])

    while len(queue) > 0:
        curr = queue.pop(0)
        next = []

        while len(curr) > 0:
            node = curr.pop(0)

            if node.left is not None:
                next.append(node.left)

            if node.right is not None:
                next.append(node.right)

        if len(next) > 0:
            queue.append(next)
            visible.append(next[len(next) - 1].val)

    return visible
